python/python/lance/dataset.py

```diff
@@ -1093,13 +1094,52 @@ def cleanup_old_versions(
             This should only be set to True if you can guarantee that no other process
             is currently working on this dataset.  Otherwise the dataset could be put
             into a corrupted state.
+
+        error_if_tagged_old_versions: bool, default True
+            Some versions may have tags associated with them. Tagged versions will
+            not be cleaned up, regardless of how old they are. If this argument is
+            set to `False`, untagged old versions will be cleaned, and tagged old
+            versions will be kept. If this argument is set to `True`, then no cleanup
+            will take place if an old version is tagged, and an exception will be
+            raised.
         """
         if older_than is None:
             older_than = timedelta(days=14)
         return self._ds.cleanup_old_versions(
-            td_to_micros(older_than), delete_unverified
+            td_to_micros(older_than), delete_unverified, error_if_tagged_old_versions
         )
 
+    def tags(self) -> dict[str, int]:
+        """
+        Return all tags in this dataset.
+        """
+        return self._ds.tags()
+
+    def checkout_tag(self, tag: str) -> "LanceDataset":
```

Comment:
What would you think about consolidating this with `checkout_version`? If `version` is `str`, then would checkout the tag; if it's an `int`, checkout the version? (Trying to avoid making `Dataset` have too many methods.)

Reply:
All of the API changes look good to me. For Python, we can use `version: int | str`.

In terms of Rust, we have a few options.

### 1. Call `checkout_tag` from Python `checkout_version`

Pro: Simple and already implemented. Non-breaking change.
Con: Requires an additional public method.

### 2. `checkout_version(&self, version: impl std::fmt::Display)`

Pro: Non-breaking change and one less public method.
Con: Requires unidiomatic coercion to string, before reverting back to `u64` if all characters are numeric.

### 3. `checkout_version(&self, version: Ref)`

```rust
enum Ref {
    Version(u64),
    Tag(&str)
}
```

Pro: Idiomatic rust and one less public method.
Con: Breaking change for `checkout_version`.

Which option would you like to go with? Or do you have any other suggestions?

Reply:
How about `checkout_version(&self, version: impl Into<Ref>)`? (compromise between 2 and 3)

Reply:
Done: 14b7efa